### PR TITLE
Fix how we do channel protection

### PR DESCRIPTION
**Pull request by ashamis**
Part of #774

### CODE REVIEW

**Comment by cimetrics** [comment]
threading_4@[4692 aka 20200207.34](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4692&view=results) vs master ewma over 30 builds from [4373](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4373&view=results) to [4656](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4656&view=results)
images

File: src/node/channels.h [hunk 1]
```diff
@@ -17,6 +17,24 @@ namespace ccf
   using SeqNo = uint64_t;
   using GcmHdr = crypto::GcmHeader<sizeof(SeqNo)>;
 
+  struct RecvNonce
+  {
+    uint8_t tid : 8;
+    uint64_t nonce : 54;
```
Comment (achamayou) [inline src/node/channels.h]: 54?

Comment (jumaffre) [inline src/node/channels.h]: Should be 56.

File: src/node/channels.h [hunk 2]
```diff
@@ -51,12 +70,18 @@ namespace ccf
         throw std::logic_error("Channel is not established for verifying");
       }
 
-      auto local_nonce = local_recv_nonce.load();
-      auto recv_nonce = header.get_iv_int();
+      auto iv = header.get_iv_int();
+      auto recv_nonce = reinterpret_cast<RecvNonce*>(&iv);
```
Comment (achamayou) [inline src/node/channels.h]: Add a ctor?

File: src/node/channels.h [hunk 3]
```diff
@@ -66,7 +91,12 @@ namespace ccf
       {
         // Set local recv nonce to received nonce only if verification is
         // successful
-        local_recv_nonce.store(recv_nonce);
+        while (local_nonce < recv_nonce->nonce)
```
Comment (jumaffre) [inline src/node/channels.h]: As discussed, this can go for now as the receiving thread is always the same for now.

File: src/node/channels.h [hunk 4]
```diff
@@ -17,6 +17,25 @@ namespace ccf
   using SeqNo = uint64_t;
   using GcmHdr = crypto::GcmHeader<sizeof(SeqNo)>;
 
+  struct RecvNonce
+  {
+    uint8_t tid : 8;
+    uint64_t nonce : 56;
```
Comment (achamayou) [inline src/node/channels.h]: sizeof(uint64_t) - sizeof(uint8_t)

**Pull request merged**